### ISSUE TITLE
Support alternate points of view in boundaries layer

### ISSUE THREAD

**Issue by nvkelso** [tilezen/vector-datasource#1552]
Natural Earth supports the concept of "points of view" (POV) to indicate the claimants for different boundary disputes. 

Right now Tilezen supports just the default "defacto" point of view, but sometimes that deviates from US State Department and United Nations views. And showing maps of India, Pakistan, China, and other areas require special treatments. 

Tilezen should support both a default POV (now) and additional POV for country and disputed boundary lines (possibly also region but Natural Earth data doesn't support that now).

**Comment by musculman** [comment]
When looking at this feature it would probably make sense to think about how to solve the fact that you would need to give different geometries depending on the `view` used.

In particular, we should not only think about a particular border being listed as belonging to two (or more) countries, but also the fact that these countries might have different understanding on where the actual border is.

Therefore, it has two main use cases:
- pure disputed borders: if `view` of Country A is used, the border of country A is here, if `view` of country B is used, the border is not here but somewhere else.
- different border depending on the `view`. As opposed to the previous case in which the border is either shown as disputed (dashed) or belonging to country A or B; in this case, the border is not there and it is somewhere else, changing the actual shape of the country.

Looking at a possible implementation for a border with an example, could be:

- `is_disputed_by`=A
- `is_accepted_by`=B
This border will be displayed in country B but not in country A.

- `is_disputed_by`=A
This border will be displayed in all views, but not in country A.

Additional attributes worth considering are:
- `claimed by` : an attribute that indicates which country claims a cartographic administrate polygon or linear administrative boundary in a disputed area.
- `controlled by`: in disputed areas which country is, according to international agreements, in control of the area.
- `line of control`: usually marks the internationally accepted boundaries, but are located in areas under dispute. Some countries might claim large territory, but de-facto control only up to this line.

**Comment by zerebubuth** [comment]
As an alternative way of looking at things; borders (national or otherwise) are largely fictitious. In some cases, there is a physical barrier marking the boundary. This makes it easy to map and draw - the big fence or wall is the border. Where there is no physical barrier, the border is by nature more abstract and porous, so it matters less exactly where it is drawn.

Some countries mandate, sometimes by law, that maps used within their jurisdiction are drawn to support their opinions (and often, ideology). Therefore, rather than calling it a `view`, we could call it a `mandatory_delusion` where it differs from the ground truth (_de facto_) feature.

The logic for drawing might look something like:

```python
ground_truth = feature.properties.get('ground_truth', 'draw')
key = "mandatory_delusion:" + this_country_code
should_draw = feature.properties.get(key, ground_truth)
if should_draw = 'draw':
  # draw it
# otherwise, don't draw it
```

In other words, a country-specific flag overrides a default "ground truth" flag, which defaults to drawing. If it turns out there are large blocks of similar delusions, we can do a two-level scheme (i.e: look for the country first and then the block for the override). This way we don't have to put a lot of extra properties on the large numbers of borders or labels which are not disputed.

**Comment by musculman** [comment]
I would really not call it something like `mandatory_delusion` or `ground_truth`. By the pure nature of this topic, it is highly sensitive and controversial, using property names like those can only cause problems. The word `view` derives from the fact that what you are seeing when using that `view` is just the world according to certain political understanding of a particular country, in other words, the political view of that country.
It's fine if another word is used, but if you look at other companies providing this solution, you will notice that the choice of words is always carefully considered.

Regarding the fact that it should not bloat non disputed borders, that is usually achieved by having a simple boolean that will tell you `is_disputed`= true|false and only if it is true you look for the additional optional values, like shown above. That way you just need a single value that can be nicely encoded.

Regarding the `ground truth`, please keep in mind, there is no such thing, only the truth according to certain political considerations, so the default view is in fact already an agreement between one or more countries. You could consider for example having the default view as the UN view, but even in that case, the UN considers many borders as disputed and there are many lines of control that further modify it, as I mentioned above. One should be very clear about what is their default view, whether it is UN, US or any other variety.

**Comment by zerebubuth** [comment]
> Regarding the `ground truth`, please keep in mind, there is no such thing, only the truth according to certain political considerations

You're right, which is why (half-jokingly, although I should have telegraphed that more clearly :bowing_man: ) I suggested we make this explicit in the property name. In many cases, there is physical infrastructure which can be measured as empirical truth, but I imagine the vast majority of borders are not marked physically in any way. I suppose that proves the point that people can be quite sensitive about being reminded of the (at best, consensual) unreality of borders.

> having a simple boolean that will tell you `is_disputed= true|false`

There's no need for an additional boolean under the scheme I described, the presence of one or more `view:${ISO_CODE}` keys implies `is_disputed=true` and the absence of any implies `false`.

**Comment by musculman** [comment]
Thanks for the additional clarification.
Let me see if I understand that proposal correctly, because I think I am missing one point there.

In the case I mentioned in my first comment, of actual changes on the country border depending on the `view`; let's consider a border that only exists for country A and let's call that border X and every other country in the world puts that border somewhere else, let's call it border Y. Remember they are different geometries.
With the system you describe, how would you encode that?
Border X would have view:iso_of_country_A and nothing more but what would Border Y have? All other possible countries? or would it simply not mention any disputed? If you do that then the `view` of country A would also show it, that is why above I mentioned this mechanism of `is_accepted_by` and `is_disputed_by`... I know it is probably confusing so that's why I am happy to see a discussion to improve it / design something else that is better. Let me know please if I misunderstood your proposal or how to code such an example.
Just for the record, it is a real life example but I did not use real country names :) :)

**Comment by zerebubuth** [comment]
I'm suggesting that it could be described as:

* Border X properties: `{"view:iso_A": "show"}`
* Border Y properties: `{"view:iso_A": "hide"}`

(If we don't like `show|hide` then we could use `accepts|disputes` or `True|False`. I think `show|hide` would make more sense to people writing the map style, but any binary pair of values would work.)

It's possible to work out `is_disputed` by asking if any key starting in `view:` is `show`. We don't need to list the countries which follow the majority view - since there are only two options, they must hold the opposite view to the minority.

So Country A will show Border X and hide Border Y, because those are the explicit overrides. Any other country will show Border Y and hide Border X, because those are the defaults.

For more complex situations, where several countries hold a different view, the scheme still works although it may be wasteful to list them all out. The maximum number of different disputants that I can see is a [six-way dispute](https://en.wikipedia.org/wiki/List_of_territorial_disputes) over [the Spratly Islands](https://en.wikipedia.org/wiki/Spratly_Islands_dispute). However, there's [very little common geometry amongst the claims](https://upload.wikimedia.org/wikipedia/commons/d/de/South_China_Sea_claims_map.jpg), so I'd assume we'd still end up with a single `view:XX` property per feature.

**Comment by musculman** [comment]
Thank you for the explanation, I had indeed misunderstood it :)
For examples like the one you listed we might need to think of having also the additional properties like  `claimed_by` or `controlled_by` depending on the routing capabilities that you want to support, if any, because you can usually ask your routing engine to avoid areas controlled by certain countries or to avoid disputed areas at all. For example if your user has a certain passport that is not allowed to cross some countries/borders.

**Comment by musculman** [comment]
I will try to create a first draft of what we have discussed on this, unless someone is already preparing it.

**Comment by nvkelso** [comment]
Suggest using "point of view" like we do language localization but instead by ISO 2-char country codes (in practice this is sparsely populated on minority of features, but then some features are very contentious!):

- **is_accepted_by:{cc}**=True
    - This border will be displayed in country B but not in country A.
    - {cc} is ISO 2-char country codes
    - This act as a "whitelist"
    - Optional value when a boundaries line is disputed in some way.
    - When value is False it is not exported in tiles
    - When no per country booleans are specified than that boundary is not disputed by anyone and renders by default.
    - When present the render only renders the line (visible=true) when it matches a stylesheet variable specifying the "point of view" as a 2-char country code.

- **is_disputed_by:{cc}**=True
    - This border will be displayed in all views, but not in country A.
    - {cc} is ISO 2-char country codes
    - This act as a "blacklist"
    - Optional value when a boundaries line is disputed in some way.
    - When value is False it is not exported in tiles
    - When no per country booleans are specified than that boundary is not disputed by anyone and renders by default.
    - When present the render will not render the line (visible=false) when it matches a stylesheet variable specifying the "point of view" as a 2-char country code.

Alternatively we could support an array of "point of view" values, like:

- **is_accepted_by**: [cc1, cc2, cc3, cc4]
- **is_disputed_by**: [cc1, cc2, cc3, cc4]
- This is not supported in current MVT encoder but there's a convention to support text string with JSON blob contents. GeoJSON and TopoJSON formats don't have issue with it.

We already have different **boundary** layer **kind** values ([list](https://mapzen.com/documentation/vector-tiles/layers/#boundary-kind-values)) for the following, and the "point of view" can be applied equally to all of them:

- country
- disputed
- line_of_control
- indefinite
- indeterminate
- lease_limit
- overlay_limit
- map_unit
- region

With respect to "line of control":

- This is already covered with kinds: `disputed`, `line_of_control`, `indefinite`, `indeterminate`, `lease_limit`, and `overlay_limit` 
- Please see how Natural Earth uses these for examples of each.
- So I think the same "claimed by" and "controlled by" can be handled the same as for normal country (and other)

**Comment by nvkelso** [comment]
I'm open to `{"point_of_view:cc":"show"}` or `{"point_of_view:cc":"hide"}`, too.

We could in that case still keep it "truthy" but make it True (show), False (hide), and Null (no opinion/use default point of view, drop the property)? 

- **"point_of_view:cc"**: `True`, `False`, `Null`

**Comment by zerebubuth** [comment]
I would prefer `{"point_of_view:cc": "show"}` or `{"point_of_view:cc": "hide"}`, because I think this is more readable than boolean values. If we really, really want to use default values, then I think we should use:

* `{"is_accepted_by:cc": True}` - means any `cc` listed in this feature accepts it, and other countries do not.
* `{"is_accepted_by:cc": False}` - means any `cc` listed in this feature _does not_ accept it, and other countries do.

If the property isn't given (i.e: `None`), then the view of the country is that of the default majority, opposite to that of the explicitly listed minority.

**Comment by musculman** [comment]
@nvkelso: I have some questions about your last comment:

* was there any issue with @zerebubuth's approach? I was preparing this as a first draft of documentation as with his solution we could save one variable by deducting the value of the others assuming the defaults. Please would you mind explaining if you found any issue with it?
* nvkelso: I am not sure how do you want to apply the `point of view` to the boundary layer kind values list you mention, do you mind claryfing that?
* you are right, we can reuse existing `line_of_control`

## Boundaries Layer

(...)

Natural Earth supports the concept of `points of view` (POV) to indicate the claimants for different boundary disputes.
When displaying a map you need to consider if you are going to support the different `points of view` or if you are just going to use the de-facto one, which is in line with US State Department and United Nations views (TODO: link?).
If a particular `point of view` is used, the boundaries of the selected country view are shown; otherwise the default boundaries view is shown.
A typical use case is that a particular area of the world is considered as *disputed* by two or more countries. Therefore, each country shows that area as part of itself. If you request a particular tile with the political view of country A, the response shows the area as belonging to A. Similarly, if you request the tile with the political view of country B, the response shows the area as belonging to B.
This means that in order to use this optional feature you need to define a `point of view` value, in a similar way as you define language localization, but using [ISO 2-char country code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2).
In each case, in order to display correctly the disputed borders you need to understand what border should be displayed in each case, and control this logic on your Rendering Engine.

_Please note_ that the the boundary layer kind `line_of_control` is also relevant for displaying correctly the `point of view`.

```
Border X properties: `{"view:iso_A": "true"}`
Border Y properties: `{"view:iso_A": "false"}`
```
Country A will show Border X and hide Border Y, because those are the explicit overrides. Any other country will show Border Y and hide Border X, because those are the defaults.
In case of borders that have different geometry depending on the `point of view` used, they will have `view:iso_C` and only the country with that `point of view`, C, in this case, will display the new geometry as everybody else will simply ignore it.
It's possible to work out what borders are disputed by asking if any key starting in `view:` is `true`. We don't need to list the countries which follow the majority view - since there are only two options, they must hold the opposite view to the minority.

#### Boundaries properties (optional):
(...)

* `view`:{cc} - it is a `boolean` field, where `true` represents `show` and `false` represents `hide`. Keep in mind that any other country not listed having this `view` will apply the default value. See the example above for more details.

#### Boundary `kind_detail` values for `country`

For `kind` `country` we can have the following optional `kind_detail` values:

* `claimed_by` - an attribute that indicates which country claims a cartographic administrate polygon or linear administrative boundary in a disputed area.
* `controlled_by` - in disputed areas which country is, according to international agreements, in control of the area.

**Comment by musculman** [comment]
@nvkelso are you still planning to do the documentation part? Otherwise I can follow with a PR on this.
But, because my latest proposal on my comment is slightly different compared to your last comment, it is perhaps better that you do it? Just let me know, thank you.

**Comment by nvkelso** [comment]
I can take care of the PR, once back in the office next week from Thanksgiving break.

**Comment by musculman** [comment]
No rush! Just wanted to know if you were going to take care of it at some point or should I go ahead.
Enjoy your days off!

**Comment by nvkelso** [comment]
I've started prototyping this out...

**Comment by nvkelso** [comment]
After prototyping this out a bit in January 2019, I now propose (v2) to "localize" the kind like `kind:iso` instead of introducing extra / duplicate spagetti lines, for the ~20 TZ languages / countries, and other major claimants (like Morocco, Pakistan, and Palestine), just like we do with `name:*` today. Meaning we should not see any significant file size increase from this approach over v1.6 data model.

This also addresses an issue where the **kind** of the feature changes per point of view, not just the visibility of the line (as was the focus of the earlier proposal above).

Similar to the last proposal, this proposal does _not_ address the issue that some labels in the **places** layer (`country` and `region` primarily) should also be disabled / changed in parallel. For example, India's `Arunachal Pradesh` region is claimed by **China** as part of the **Tibet** `region` and ideally there would be a way to mark that place's localized `kind:*` to `unrecognized` as well, but that's out of scope for v1.7 (though consistent with this proposal).

## Prototype data for Natural Earth v5.0.0-pre1:

- [ne_v5.0.0-pre1-boundaries-pov.zip](https://github.com/tilezen/vector-datasource/files/2811297/ne_v5.0.0-pre1-boundaries-pov.zip)
- Three boundary line files
    - ne_10m_admin_0_boundary_lines_land.shp
    - ne_10m_admin_0_boundary_lines_map_units.shp
    - **ne_10m_admin_0_boundary_lines_disputed_areas.shp** (new)
    - _I'm ignoring the ne_50m_admin_0_boundary_lines_land.shp and ne_50m_admin_0_boundary_lines_disputed_areas.shp files in this prototype._

## Tilezen considerations:

- We currently [set TZ `kind` values](https://github.com/tilezen/vector-datasource/blob/master/yaml/boundaries.yaml#L97-L174) from the NE `featurecla` (`featureclass` or `fclass`) property
- We [load data to the data assets S3 ZIP](https://github.com/tilezen/vector-datasource/blob/master/data/assets.yaml#L92-L106), but currently ignore the `ne_10m_admin_0_boundary_lines_disputed_areas` file.
- We should start loading the `ne_10m_admin_0_boundary_lines_disputed_areas` file, with 
    - Kind value mapping as:
        - `Breakaway` (NE) > `disputed_breakaway` (TZ)
        - `Claim boundary` > `disputed_claim`
        - `Elusive frontier` > `disputed_elusive`
        - `Reference line` > `disputed_reference_line`
- The `kind_detail` should be **2** for the new `disputed_*` kinds.
- We'll need to add support for the following localized `kind:*` values... and remap their values from the raw list to the YAML list of TZ kind values, perhaps with a Python transform (or in the YAML like the `ne_region_boundaries_min_zoom` lookup)? See example below.
- The localized `kind:*` values includes a special `Unrecognized` (map to lowercase `unrecognized`) value that is used to "disable" the visibility of a line. It's not present as a basic `kind` value.
- If the localized `kind:*` value is null, that property should be dropped.
- The `min_zoom` should come from the Natural Earth min_zoom value, not default to 0. There are many disputed boundaries that should only be visible at later zooms, and that has been curated.

```
  - &ne_country_boundaries_kind
    lookup:
      key: { col: featurecla }
      op: '=='
      table:
        - [ 'Disputed (please verify)', 'disputed' ]
        - [ 'Indefinite (please verify)', 'indefinite' ]
        - [ 'Indeterminant frontier', 'indeterminate' ]
        - [ 'International boundary (verify)', 'country' ]
        - [ 'Lease limit', 'lease_limit' ]
        - [ 'Line of control (please verify)', 'line_of_control' ]
        - [ 'Overlay limit', 'overlay_limit' ]
        - [ 'Unrecognized', 'unrecognized' ]
        - [ 'Map unit boundary', 'map_unit' ]
        - [ 'Breakaway', 'disputed_breakaway' ]
        - [ 'Claim boundary', 'disputed_claim' ]
        - [ 'Elusive frontier', 'disputed_elusive' ]
        - [ 'Reference line', 'disputed_reference_line' ]
      default: null
```
```
global:
  - &ne_localized_kind_properties
    'kind:iso': {col: FCLASS_ISO }
    'kind:us': {col: FCLASS_US }
    'kind:fr': {col: FCLASS_FR }
    'kind:ru': {col: FCLASS_RU }
    'kind:es': {col: FCLASS_ES }
    'kind:cn': {col: FCLASS_CN }
    'kind:tw': {col: FCLASS_TW }
    'kind:in': {col: FCLASS_IN }
    'kind:np': {col: FCLASS_NP }
    'kind:pk': {col: FCLASS_PK }
    'kind:de': {col: FCLASS_DE }
    'kind:gb': {col: FCLASS_GB }
    'kind:br': {col: FCLASS_BR }
    'kind:il': {col: FCLASS_IL }
    'kind:ps': {col: FCLASS_PS }
    'kind:sa': {col: FCLASS_SA }
    'kind:eg': {col: FCLASS_EG }
    'kind:ma': {col: FCLASS_MA }
    'kind:pt': {col: FCLASS_PT }
    'kind:ar': {col: FCLASS_AR }
    'kind:jp': {col: FCLASS_JP }
    'kind:ko': {col: FCLASS_KO }
    'kind:vn': {col: FCLASS_VN }
    'kind:tr': {col: FCLASS_TR }
    'kind:id': {col: FCLASS_ID }
    'kind:pl': {col: FCLASS_PL }
    'kind:gr': {col: FCLASS_GR }
    'kind:it': {col: FCLASS_IT }
    'kind:nl': {col: FCLASS_NL }
    'kind:se': {col: FCLASS_SE }
    'kind:bd': {col: FCLASS_BD }
```
## Example POV: 

_In QGIS these are accomplished by a `coalesce( "FCLASS_CN", "FEATURECLA" )` to prefer the localized kind, but fallback to the default kind on all the 3 boundaries SHP themes. The polygons are also built using similar attributes and new build targets – but since we don't include admin polygons in Tilezen those are not included in the prototype data archive above._

China:
<img width="1199" alt="screen shot 2019-01-29 at 22 46 09" src="https://user-images.githubusercontent.com/853051/51964481-8d02d200-241b-11e9-94ae-48d8b90f3423.png">

India:
<img width="1200" alt="screen shot 2019-01-29 at 22 46 20" src="https://user-images.githubusercontent.com/853051/51964483-8d02d200-241b-11e9-9190-0249c31f7d11.png">

Pakistan:
<img width="1197" alt="screen shot 2019-01-29 at 22 46 31" src="https://user-images.githubusercontent.com/853051/51964484-8d02d200-241b-11e9-9d79-d81d9a40454a.png">

Default:
<img width="1199" alt="screen shot 2019-01-29 at 22 46 42" src="https://user-images.githubusercontent.com/853051/51964485-8d02d200-241b-11e9-9dd0-9e9dce3cd278.png">

## Example data table attributes

_See `featurecla` column at far left compared to `fclass_cn` column at far right._

<img width="1479" alt="screen shot 2019-01-29 at 23 00 22" src="https://user-images.githubusercontent.com/853051/51965102-5af26f80-241d-11e9-9df3-999789c5d5c0.png">
<img width="1407" alt="screen shot 2019-01-29 at 22 58 53" src="https://user-images.githubusercontent.com/853051/51965103-5af26f80-241d-11e9-8224-0a9621049877.png">

**Comment by nvkelso** [comment]
Looks like we're sometimes missing the POV kind values, here for China:

![image](https://user-images.githubusercontent.com/853051/52768839-692ab900-2fe3-11e9-9fd9-5cb75b4cc09c.png)

Versus what was in the shapefile delivery from Natural Earth v5.0.0-pre1:

![image](https://user-images.githubusercontent.com/853051/52768818-5adc9d00-2fe3-11e9-81ba-2a80e6490fc9.png)

@zerebubuth can you have a look, please?

**Comment by nvkelso** [comment]
But otherwise this is working as expected :)

There's also work to add this in from the Natural Earth 1:50M and 1:110M scale sets... and from populated_places and states_provinces as some country and region capitals aren't recognized, and some regions aren't recognized. That data still needs to be created.

<img width="1236" alt="screen shot 2019-02-13 at 23 04 36" src="https://user-images.githubusercontent.com/853051/52769113-2e755080-2fe4-11e9-96b0-3b74a8a24c69.png">
<img width="1202" alt="screen shot 2019-02-13 at 23 04 24" src="https://user-images.githubusercontent.com/853051/52769115-2e755080-2fe4-11e9-95e6-be5e97b60c75.png">
<img width="1221" alt="screen shot 2019-02-13 at 23 04 16" src="https://user-images.githubusercontent.com/853051/52769116-2e755080-2fe4-11e9-84cb-7558b0f73ff2.png">

**Comment by nvkelso** [comment]
BTW, the missing China boundary at zoom 6 along Nepal is because it's NE min_zoom of 7, so that'll need to be changed in the source data.

**Comment by zerebubuth** [comment]
> Looks like we're sometimes missing the POV kind values, here for China:

I think you might have been hovering over this line instead:

![image](https://user-images.githubusercontent.com/271360/52787470-d4ad6e80-3055-11e9-8922-d6c01fa3cc7c.png)

The version of the shapefile I have here says its default feature class is "Indefinite (please verify)" and only override is `FCLASS_IN` to "Unrecognized". The line to the west, which I think is the one you selected in QGIS, looks like this in tiles to me:

<img width="454" alt="screen shot 2019-02-14 at 12 35 45" src="https://user-images.githubusercontent.com/271360/52787718-9795ac00-3056-11e9-9f73-fda3035bc117.png">

**Comment by nvkelso** [comment]
Ah, you're right! I'll update the boundary attributes upstream in Natural Earth to fill in the missing info. Thanks :)

**Comment by nvkelso** [comment]
This is working as expected! 

For final build I'll provide v5.0.0-pre2 files which address Cyprus, Israel, and few other bug fixes.

Followup in https://github.com/tilezen/vector-datasource/issues/1840 for lower zooms, and the places layer in v1.8.

**Comment by nvkelso** [comment]
Fixed via https://github.com/tilezen/vector-datasource/pull/1809.
